ext/twig/twig.c

```diff
@@ -0,0 +1,758 @@
+static zval twig_call_user_func_array(zval *object, char *function, zval *arguments)
+{
+	zend_fcall_info fci = {0};
+	zval retval;
+
+	zval zfunction;
+
+	zend_fcall_info_args(&fci, arguments);
+
+	ZVAL_STRING(&zfunction, function);
+
+	fci.size           = sizeof(fci);
+#if IS_PHP_70
+	fci.function_table = EG(function_table);
+	fci.symbol_table   = NULL;
+#endif
+	fci.function_name  = zfunction;
+	fci.object         = Z_OBJ_P(object);
+	fci.retval         = &retval;
+	fci.no_separation  = 0;
+
+	if (zend_call_function(&fci, NULL) == FAILURE) {
+		ZVAL_FALSE(&retval);
+	}
+
+	if (fci.params) {
+		efree(fci.params);
+	}
+
+	zval_ptr_dtor(&zfunction);
+
+	return retval;
+}
+
+static int twig_call_boolean(zval *object, char *functionName)
+{
+	zval ret;
+
+	ret = twig_call_user_func_array(object, functionName, NULL);
+
+	return i_zend_is_true(&ret);
+}
+
+static zval *twig_get_array_element_zval(zval *class, zval *prop_name)
+{
+	if (Z_TYPE_P(class) == IS_ARRAY) {
+		switch(Z_TYPE_P(prop_name)) {
+			case IS_NULL:
+				return zend_hash_find(HASH_OF(class), CG(empty_string));
+			case IS_TRUE:
+				return zend_hash_index_find(HASH_OF(class), 1);
+			case IS_FALSE:
+				return zend_hash_index_find(HASH_OF(class), 0);
+			case IS_DOUBLE:
+				return zend_hash_index_find(HASH_OF(class), zend_dval_to_lval(Z_DVAL_P(prop_name)));
+			case IS_LONG:
+				return zend_hash_index_find(HASH_OF(class), Z_LVAL_P(prop_name));
+			case IS_STRING:
+				return zend_symtable_find(HASH_OF(class), Z_STR_P(prop_name));
+			default:
+				return NULL;
+		}
+	}
+
+	return twig_get_arrayobject_element(class, prop_name);
+}
+
+static zval *twig_read_property(zval *object, zval *propname)
+{
+	zval *tmp = NULL;
+	zend_class_entry *cur_scope = zend_get_executed_scope();
+
+	if (Z_OBJ_HT_P(object)->read_property) {
+		if (!cur_scope) {
+			EXECUTOR_SCOPE = Z_OBJCE_P(object);
+		}
+		tmp = Z_OBJ_HT_P(object)->read_property(object, propname, BP_VAR_IS, NULL, &nil);
+		if (!cur_scope) {
+			EXECUTOR_SCOPE = NULL;
+		}
+	}
+	return tmp;
+}
+
+static int twig_has_property(zval *object, zval *propname, zend_string *prop)
+{
+	int retval = 0;
+
+	if (Z_OBJ_HT_P(object)->has_property) {
+		retval = Z_OBJ_HT_P(object)->has_property(object, propname, 0, NULL);
+	}
+	if ((retval == 0) && (Z_OBJ_HT_P(object)->get_properties)) {
+		return zend_hash_exists( Z_OBJ_HT_P(object)->get_properties(object), prop);
+	}
+
+	return retval;
+}
+
+static int twig_call_method(zval *object, char *method /*lowercased */, zval *result, zval *arg1, zval *arg2)
+{
+	int success;
+	zval retval;
+
+	if (arg2) {
+		zend_call_method(object, Z_OBJCE_P(object), NULL, method, strlen(method), &retval, 2, arg1, arg2);
+	} else {
+		zend_call_method(object, Z_OBJCE_P(object), NULL, method, strlen(method), &retval, 1, arg1, NULL);
+	}
+
+	success = Z_TYPE(retval) == IS_TRUE;
+
+	if (result) {
+		ZVAL_COPY(result, &retval);
+	}
+
+	zval_ptr_dtor(&retval);
+
+	return success;
+}
+
+static int twig_has_extension(zval *obj)
```

Comment:
would be more readable if it was named ``twig_has_sandbox_extension``, as it checks for the sandbox extension, not for any extension (the macro should be renamed too)